src/ini.cr

```diff
@@ -1,22 +1,47 @@
 class INI
+  # Exception thrown on an INI parse error.
+  class ParseException < Exception
+    getter line_number : Int32
+    getter column_number : Int32
+
+    def initialize(message, @line_number, @column_number)
+      super "#{message} at #{@line_number}:#{@column_number}"
+    end
+
+    def location
+      {line_number, column_number}
+    end
+  end
+
   # Parses INI-style configuration from the given string.
+  # Raises a `ParseException` on any errors.
   #
   # ```
   # INI.parse("[foo]\na = 1") # => {"foo" => {"a" => "1"}}
   # ```
   def self.parse(str) : Hash(String, Hash(String, String))
     ini = {} of String => Hash(String, String)
-
+    lines = str.lines.map(&.lstrip)
     section = ""
-    str.each_line do |line|
-      if line =~ /\s*(.*[^\s])\s*=\s*(.*[^\s])/
-        ini[section] ||= {} of String => String if section == ""
-        ini[section][$1] = $2
-      elsif line =~ /\[(.*)\]/
-        section = $1
+
+    lines.each_with_index(1) do |line, lineno|
+      next if line.empty?
```

Comment:
This could all be a `case` expression to make the code better understandable:
```
case line[0]?
when nil, '#', ';'
  next
when '['
  ...
else
  ...
end
```

Reply:
I'd prefer the empty case to be outside the `case`, but apart from that, yes.